File: src/Monad/Fuunctor.scala
package Monad

import Intro._
import Structure._

trait Fuunctor[F[_]] {
  def fmaap[A, B](f: A => B): F[A] => F[B]

  // Provided as a synonym for `fmaap` but with the arguments flipped.
  // This helps the type-inferencer.
  final def apply[A, B](a: F[A])(f: A => B): F[B] =
    fmaap(f)(a)
}

object Fuunctor {
  // Exercise 1
  // Relative Difficulty: 1
  implicit val IdFuunctor: Fuunctor[Id] =
    new Fuunctor[Id] {
      def fmaap[A, B](f: A => B) =
        sys.error("todo")
    }

  // Exercise 2
  // Relative Difficulty: 2
  implicit val ListFuunctor: Fuunctor[List] =
    new Fuunctor[List] {
      def fmaap[A, B](f: A => B) =
        sys.error("todo")
    }

  // Exercise 3
  // Relative Difficulty: 2
  implicit val OptionalFuunctor: Fuunctor[Optional] =
    new Fuunctor[Optional] {
      def fmaap[A, B](f: A => B) =
        sys.error("todo")
    }

  // Exercise 4
  // Relative Difficulty: 3
  implicit def Function1Fuunctor[T]: Fuunctor[({type l[a] = T => a})#l] =
    new Fuunctor[({type l[a] = T => a})#l] {
      def fmaap[A, B](f: A => B) =
        sys.error("todo")
    }

  ///////////////////////
  // SUPPORT LIBRARIES //
  ///////////////////////

  implicit val ScalaListFuunctor: Fuunctor[scala.List] =
    new Fuunctor[scala.List] {
      def fmaap[A, B](f: A => B) =
        _ map f
    }

}

File: src/Monad/Fuunctor.hs
module Monad.Fuunctor where

import Intro.Id
import Intro.Optional
import Intro.Validation
import Structure.List

class Fuunctor f where
  fmaap :: (a -> b) -> f a -> f b

-- Exercise 1
-- Relative Difficulty: 1
--
-- | Maps a function on the Id functor.
--
-- >>> fmaap (+1) (Id 2)
-- Id 3
instance Fuunctor Id where
  fmaap =
    mapId

-- Exercise 2
-- Relative Difficulty: 2
--
-- | Maps a function on the List functor.
--
-- >>> fmaap (+1) Nil
-- []
--
-- >>> fmaap (+1) (1 :. 2 :. 3 :. Nil)
-- [2,3,4]
instance Fuunctor List where
  fmaap =
    maap

-- Exercise 3
-- Relative Difficulty: 2
--
-- | Maps a function on the Optional functor.
--
-- >>> fmaap (+1) Empty
-- Empty
--
-- >>> fmaap (+1) (Full 2)
-- Full 3
instance Fuunctor Optional where
  fmaap =
    mapOptional

-- Exercise 4
-- Relative Difficulty: 3
--
-- | Maps a function on the reader ((->) t) functor.
--
-- >>> fmaap (+1) (*2) 8
-- 17
instance Fuunctor ((->) t) where
  fmaap f g =
    \x -> f (g x)

-- Exercise 5
-- Relative Difficulty: 2
--
-- | Maps a function on an IO program.
--
-- >>> fmaap reverse (putStr "hi" >> return "abc")
-- hi"cba"
instance Fuunctor IO where
  fmaap =
    fmap

-----------------------
-- SUPPORT LIBRARIES --
-----------------------

instance Fuunctor [] where
  fmaap =
    fmap
